### ISSUE TITLE
Dynamic group mirroring the file system.

### ISSUE THREAD

**Issue by koppor** [JabRef/jabref#10930]
As user, I organize my PDFs in folders. Example:

```
Related_Work_Collection
├── Quantum_Computing
│   ├── Algorithms
│   │   ├── Quantum_Factorization_Overview.pdf
│   │   └── Grover_Search_Implementation.pdf
│   ├── Error_Correction
│   │   ├── Topological_Qubits_and_Error_Correction.pdf
│   │   └── Quantum_Codec_Principles.pdf
│   └── Hardware
│       └── Superconducting_Qubits_Advancements.pdf
├── Machine_Learning
│   ├── Neural_Networks
│   │   ├── Deep_Learning_Architectures.pdf
│   │   └── CNN_for_Image_Recognition.pdf
│   ├── Reinforcement_Learning
│   │   └── RL_in_Video_Games.pdf
│   └── Natural_Language_Processing
│       ├── BERT_and_Transformers.pdf
│       └── Sentiment_Analysis_Techniques.pdf
└── Environmental_Science
    ├── Climate_Change
    │   └── Impact_of_Climate_Change_on_Biodiversity.pdf
    ├── Renewable_Energy
    │   ├── Solar_Energy_Trends.pdf
    │   └── Wind_Power_Innovations.pdf
    └── Conservation
        ├── Wildlife_Conservation_Strategies.pdf
        └── Marine_Protection_Efforts.pdf
```

When one wants to have this hierarchy in JabRef, one needs to **manually** create groups and sub groups.

A new type of group "Directory structure" should be created:

![image](https://github.com/JabRef/jabref/assets/1366654/7ab4fc43-2a67-4e1c-9fc7-bdd669471d66)

JabRef then mirrors the directory structure in the group tree. Sub groups are automatically created.

Currently: Only inbound mirroring. Use [WatchService](https://www.baeldung.com/java-nio2-watchservice) to watch for changes.

All BibEntries having a file in the selected directory belong to that group.

File added to a folder -> automatically added to the respective group and a new BibEntry created (with PDF import)

Implementation hint:

- A user selects the root directory, all sub groups are automatically updated (created and remved) by JabRef upon file system changes.
- Implement a `DirectoryUpdateMonitor` (similar to org.jabref.model.util.FileUpdateMonitor) which watches directories and sends updates to the group tree (see third bullet point at https://github.com/JabRef/jabref/pull/10937#issuecomment-1975224043 in another context)

---

⚠️ Do think of handling the case when a group with the same directory name already exists. ⚠️

---

Future work: 

1.  Group moved -> On file system also renames and BibTeX entry "files" field adapted accordingly
2. Group renamed -> On file system also renamed and BibTeX entry "files" field adapted accordingly
3.  file moved -> entry assigned to another group BibTeX entry "files" field adapted accordingly

---

Related issues:

- https://github.com/koppor/jabref/issues/501
- https://github.com/koppor/jabref/issues/502

**Comment by Siedlerchr** [comment]
The other way round is working: 

> With file directory pattern, JabRef can automatically create subfolders and move the files into the directory based on the defined pattern. As an example, you have a single folder, e.g. papers for all your PDFs linked to their corresponding entry in JabRef. Now you want to arrange them according to defined groups. Let's say you have two groups, Automation and Biology, with a couple of entries.
Now set the file directory pattern to: [groups:(unknown)]

https://docs.jabref.org/finding-sorting-and-cleaning-entries/filelinks#filename-format-and-file-directory-pattern

**Comment by eggestig** [comment]
Hi! We are a group of 5 uni students who'd like to try and tackle this issue. We would love to be assigned to this issue and if there's anything extra to help us get started that would be much appreciated as well!

**Comment by upfrog** [comment]
Howdy! I'd love to give this issue a go. Could I please be assigned to it?

**Comment by upfrog** [comment]
Howdy - I just wanted to document that I haven't abandoned this. I foolishly grabbed this issue before a bunch of travelling, but that's all over now.

**Comment by koppor** [comment]
@upfrog Any updates?

**Comment by jpbochicchio** [comment]
Hello - I would like to take this one. Can you please assign it to me?

Thanks!

**Comment by InAnYan** [comment]
A crazy observation:

In theory, if this feature will be implemented, JabRef could become a file-manager 😆  (though, additional machinery is needed for that: 1) represent files as bib -- this is easy, 2) represent folders as groups -- will be implemented in this issue.

**Comment by koppor** [comment]
> In theory, if this feature will be implemented, JabRef could become a file-manager 😆 

Yeah - and hopefully more knowledge-centered than [TagSpaces](https://www.tagspaces.org/).

**Comment by Julien384760** [comment]
Hi ! As you can see, a friend of mine is currently working on this issue, is it possible to unassign me and to assign him instead so that he can do a draft PR ?

**Comment by koppor** [comment]
> Hi ! As you can see, a friend of mine is currently working on this issue, is it possible to unassign me and to assign him instead so that he can do a draft PR ?

Your friend neesd to comment on this PR in order to get assigned.

**Comment by NiMv1** [comment]
I'm working on this issue in PR #14668
